### ISSUE TITLE
VRAM burst in 1.1.5

### ISSUE THREAD

**Issue by chungexcy** [lltcggie/waifu2x-caffe#38]
1.1.5 may request several times of VRAM when algorithm benchmark runs.

![vram](https://cloud.githubusercontent.com/assets/12037037/16706482/43c5e012-4563-11e6-8422-beaac84359a0.png)

This is a log of my VRAM usage. Here I show 6 operation with different crop_size.

Stage 1: crop_size = 576
Stage 2: crop_size = 288
Stage 3: crop_size = 384
Stage 4: crop_size = 576

I close waifu2x-caffe-gui and then open it.
Stage 5: crop_size = 384
Stage 6: crop_size = 576

In Stage 2/3, there is a burst of VRAM usage, but not always happens.

If crop_size = 576 and the usage goes beyong 4GB of my limit, the error log is 

```
Log file created at: 2016/07/08 23:10:26
Running on machine: SERVER_USAGE
Log line format: [IWEF]mmdd hh:mm:ss.uuuuuu threadid file:line] msg
F0708 23:10:26.411075  6272 cudnn_conv_layer.cpp:191] Check failed: status == CUDNN_STATUS_SUCCESS (4 vs. 0)  CUDNN_STATUS_INTERNAL_ERROR
```

1.1.4 doesn't have this bug.

**Comment by lltcggie** [comment]
ver 1.1.6で直ったと思います。 https://github.com/lltcggie/caffe/commit/3132bce11647052b997bab6445b0cf2888b01664
試してみてください。

**Comment by chungexcy** [comment]
In 1.16 ver., this bug is still there.
When crop_size = 720 or 540 or even smaller 240, the VRAM is fully used (4GB) in benchmark and then drops back to the normal level.
But when crop_size = 576, this number seems trigger the bug.

I test `cudnnFindConvolutionForwardAlgorithm()` on ubuntu caffe. I didn't see a huge VRAM burst and past crop-size=1024.

**Comment by lltcggie** [comment]
強制終了は起こりましたか？
ver 1.1.6ではVRAMの最大容量を超えないように実装を変更しました。
なので、一時的にギリギリまでVRAMが使われるのは仕様です。

**Comment by chungexcy** [comment]
Yes, force stop happened. in 1.1.6. Same error reason in the log file.
I think forced stop caused by the exceeding the maximum VRAM is not the fundamental reason. Is it possible to figure out why benchmark might use extra VRAM sometimes?

**Comment by lltcggie** [comment]
なるほど、まだ強制終了が発生するんですね。
cudnnFindConvolutionForwardAlgorithmEx()を使用することで、cuDNN内で使われるVRAMの量を完全に制御したつもりだったんですけどね...
今はフリーのVRAMの最大9割をワークスペース用に確保するようにしているので、最低限200MBは残すようにしてみます。

**Comment by lltcggie** [comment]
なるほど、まだ強制終了が発生するんですね。
cudnnFindConvolutionForwardAlgorithmEx()を使用することで、cuDNN内で使われるVRAMの量を完全に制御したつもりだったんですけどね...
今はフリーのVRAMの最大9割をワークスペース用に確保するようにしているので、最低限200MBは残すようにしてみます。
